tests/modules/conditional_random_field_test.py

```diff
@@ -138,3 +140,59 @@ def test_viterbi_tags(self):
             most_likely_tags.append(list(most_likely))
 
         assert viterbi_tags == most_likely_tags
+
+    def test_constrained_viterbi_tags(self):
+        constraints = {(0, 0), (0, 1),
+                       (1, 1), (1, 2),
+                       (2, 2), (2, 3),
+                       (3, 3), (3, 4),
+                       (4, 4), (4, 0)}
+
+        crf = ConditionalRandomField(num_tags=5, constraints=constraints)
+        crf.transitions = torch.nn.Parameter(self.transitions)
+        crf.start_transitions = torch.nn.Parameter(self.transitions_from_start)
+        crf.end_transitions = torch.nn.Parameter(self.transitions_to_end)
+
+        mask = Variable(torch.LongTensor([
+                [1, 1, 1],
+                [1, 1, 0]
+        ]))
+
+        viterbi_tags = crf.viterbi_tags(self.logits, mask)
+
+        # Now the tags should respect the constraints
+        assert viterbi_tags == [
+                [2, 3, 3],
+                [2, 3]
+        ]
+
+    def test_allowed_transitions(self):
+        # pylint: disable=bad-whitespace,bad-continuation
+        bio_labels = ['O', 'B-X', 'I-X', 'B-Y', 'I-Y']
+        #              0     1      2      3      4
+        allowed = allowed_transitions("BIO", dict(enumerate(bio_labels)))
+        assert set(allowed) == {
```

Comment:
Can you just add a comment here and below that empty spaces in this beautifully laid out matrix are not allowed - it took me 2 minutes of starring at this thinking you had lost your mind before I understood what was happening

Reply:
done